### PR TITLE
Switch to hash table from AVL. Lock free in the presence of epoch allocator.

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Switch to hash table from AVL. Lock-free in the presence of epoch allocator.

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/platform/ebpf_hash_table.c [hunk 1]
```diff
@@ -3,52 +3,175 @@
 
 #include "ebpf_platform.h"
 
+typedef struct _ebpf_hash_bucket_header
```
Comment (shankarseal) [inline libs/platform/ebpf_hash_table.c:6]: suggest: rename bucket_entry ?

File: libs/platform/ebpf_hash_table.c [hunk 2]
```diff
@@ -111,125 +242,227 @@ ebpf_result_t
...
-        goto Done;
+    // Perform an atomic insert of the key if it's not present or replace it if
+    // if the key is already present.
+    // Note: This assumes either the table is synchronized by an external lock
+    // or that memory is freed when the epoch ends.
+    for (;;) {
+        // Capture the value of the list.
+        previous_bucket = &(hash_table->buckets[hash % hash_table->bucket_count]);
+        first_bucket = hash_table->buckets[hash % hash_table->bucket_count];
+
```
Comment (shankarseal) [inline libs/platform/ebpf_hash_table.c:313]: previous_bucket = &first_bucket - will make the intention slightly clearer IMO.

File: libs/execution_context/ebpf_program.c [hunk 3]
```diff
@@ -249,6 +249,9 @@ ebpf_program_create(ebpf_program_t** program)
         goto Done;
     }
 
+    ebpf_list_initialize(&local_program->links);
+    ebpf_lock_create(&local_program->links_lock);
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:253]: Where is this used?  Should this be part of a different PR?

File: libs/platform/ebpf_hash_table.c [hunk 4]
```diff
@@ -3,52 +3,175 @@
...
-// AVL tree gives a single struct containing both key and value.
-// Compare can be called with a partial struct only containing the key.
-// Do not access beyond map->ebpf_map_definition.key_size bytes.
-static RTL_GENERIC_COMPARE_RESULTS
-_ebpf_hash_map_compare(_In_ struct _RTL_AVL_TABLE* avl_table, _In_ void* first_struct, _In_ void* second_struct)
+/**
+ * @brief Perform a rotate left on a value.
+ *
+ * @param value Value to be rotated.
+ * @param count Count of bits to rotate.
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:29]:
```suggestion
 * @param[in] value Value to be rotated.
 * @param[in] count Count of bits to rotate.
```

File: libs/platform/ebpf_hash_table.c [hunk 5]
```diff
@@ -3,52 +3,175 @@
...
+ * @return Rotated value.
+ */
+static inline unsigned long
+_ebpf_rol(uint32_t value, size_t count)
+{
+    return (value << count) | (value >> (32 - count));
+}
+
+// Ported from https://github.com/aappleby/smhasher
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:38]: That repo has no license listed.  What license is this code based on?

File: libs/platform/user/ebpf_platform_user.cpp [hunk 6]
```diff
@@ -263,6 +198,20 @@ ebpf_interlocked_compare_exchange_int32(_Inout_ volatile int32_t* destination, i
...
 }
 
+void*
+ebpf_interlocked_compare_exchange_pointer(_Inout_ void* volatile* destination, void* exchange, void* comperand)
+{
+    return InterlockedCompareExchangePointer((void* volatile*)destination, exchange, comperand);
+}
+
+uint32_t
+ebpf_random_uint32()
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:208]: There's a helper we need (issue #342) called
`u32 bpf_get_prandom_u32(void)`.
Should this be renamed, or should that helper just call this API?

File: tools/port_quota/port_quota.cpp [hunk 7]
```diff
@@ -73,10 +74,17 @@ load(int argc, char** argv)
     }
 
     error = bpf_link__pin(link, program_link);
+    if (error != ERROR_SUCCESS) {
+        fprintf(stderr, "Failed to pin eBPF link: %d\n", result);
```
Comment (dthaler) [inline tools/port_quota/port_quota.cpp:78]: Looks like a bug
```suggestion
        fprintf(stderr, "Failed to pin eBPF link: %d\n", error);
```

File: tools/port_quota/port_quota.cpp [hunk 8]
```diff
@@ -73,10 +74,17 @@ load(int argc, char** argv)
...
+        return 1;
+    }
+
+    error = bpf_program__pin(program, program_path);
     if (error != ERROR_SUCCESS) {
         fprintf(stderr, "Failed to pin eBPF program: %d\n", result);
```
Comment (dthaler) [inline tools/port_quota/port_quota.cpp:84]:
```suggestion
        fprintf(stderr, "Failed to pin eBPF program: %d\n", error);
```

File: libs/platform/ebpf_hash_table.c (hunk 5 above)
Comment (Alan-Jowett) [reply libs/platform/ebpf_hash_table.c:38]: It is a re-implementation based on the algorithm:
https://github.com/aappleby/smhasher/blob/61a0530f28277f2e850bfc39600ce61d02b518de/src/MurmurHash3.cpp#L2

License says public domain.

File: libs/execution_context/ebpf_program.c (hunk 3 above)
Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:253]: Driver verifier bug checks on this. I can move it to a different PR if needed, but it was blocking testing of this change (I ran with driver verifier enabled to catch bugs in my change and this was hit).

File: libs/platform/ebpf_hash_table.c (hunk 5 above)
Comment (dthaler) [reply libs/platform/ebpf_hash_table.c:38]: Great, would be nice to say so in the // comment.

File: libs/execution_context/ebpf_program.c (hunk 3 above)
Comment (dthaler) [reply libs/execution_context/ebpf_program.c:253]: I didn't see anything in the PR description about this.  So either move to another PR or update the PR description here.   Normally we try to avoid unrelated things being in the same PR, but not a hard rule, so up to you.

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c:253]: Agreed. https://github.com/microsoft/ebpf-for-windows/pull/395

File: libs/platform/ebpf_hash_table.c [hunk 9]
```diff
@@ -3,52 +3,175 @@
...
 
-static void
-_ebpf_hash_table_free(struct _RTL_AVL_TABLE* avl_table, void* buffer)
+/**
+ * @brief Accepts a chain of buckets and returns pointer to a bucket with
+ * a matching key or returns NULL of not found.
+ *
+ * @param[in] hash_table Hash-table this bucket belongs to.
+ * @param[in,out] previous_bucket Pointer to the start of the bucket chain on
+ * entry. Pointer to the last bucket before the bucket containing the key on
```
Comment (shankarseal) [inline libs/platform/ebpf_hash_table.c:153]: Please re-write. On return it is pointing to the next field of the previous value of the matched value (if found) or next field of the last element (if not found).

File: libs/execution_context/ebpf_program.c (hunk 3 above)
Comment (saxena-anurag) [reply libs/execution_context/ebpf_program.c:253]: This is being initialized in ebpf_program_initialize()

File: libs/platform/ebpf_hash_table.c [hunk 10]
```diff
@@ -111,125 +242,227 @@ ebpf_result_t
...
+        bucket = _ebpf_hash_table_find_bucket(hash_table, &previous_bucket, key);
+
+        if (bucket == NULL) {
+            retval = EBPF_KEY_NOT_FOUND;
+            goto Done;
+        }
+
+        // Set the previous bucket's next value to bucket next value only if
+        // previous bucket's next value has not changed.
+        if (ebpf_interlocked_compare_exchange_pointer((void* volatile*)previous_bucket, bucket->next, bucket) !=
```
Comment (shankarseal) [inline libs/platform/ebpf_hash_table.c:377]: Alan will check for correctness here. Likely a bug if a->b->c->d (b,c) are getting deleted concurrently.

File: libs/platform/ebpf_hash_table.c [hunk 11]
```diff
@@ -3,52 +3,175 @@
...
-    struct _RTL_AVL_TABLE avl_table;
-    ebpf_hash_table_compare_result_t (*compare_function)(const uint8_t* key1, const uint8_t* key2);
+    uint32_t bucket_count;
+    uint32_t entry_count;
+    uint32_t seed;
     size_t key_size;
     size_t value_size;
     void* (*allocate)(size_t size);
     void (*free)(void* memory);
+    void (*extract)(const uint8_t* value, const uint8_t** data, size_t* num);
```
Comment (shankarseal) [inline libs/platform/ebpf_hash_table.c:21]: prefer separate compare_cb and optional hash_cb. If hash_cb == NULL, use murmur.

File: libs/platform/ebpf_pinning_table.c [hunk 12]
```diff
@@ -5,31 +5,20 @@
...
 typedef struct _ebpf_pinning_table
 {
     _Requires_lock_held_(&lock) ebpf_hash_table_t* hash_table;
     ebpf_lock_t lock;
 } ebpf_pinning_table_t;
 
-static ebpf_hash_table_compare_result_t
-_ebpf_pining_table_compare_function(const uint8_t* key1, const uint8_t* key2)
+static void
+_ebpf_pining_table_extract(const uint8_t* value, const uint8_t** data, size_t* length)
```
Comment (dthaler) [inline libs/platform/ebpf_pinning_table.c:17]:
```suggestion
_ebpf_pinning_table_extract(_In_ const uint8_t* value, _Outptr_ const uint8_t** data, _Out_ size_t* length)
```

File: libs/platform/ebpf_pinning_table.c [hunk 13]
```diff
@@ -63,7 +52,8 @@ ebpf_pinning_table_allocate(ebpf_pinning_table_t** pinning_table)
         ebpf_free,
         sizeof(ebpf_utf8_string_t*),
         sizeof(ebpf_pinning_entry_t*),
-        _ebpf_pining_table_compare_function);
+        EBPF_PINNING_TABLE_BUCKET_COUNT,
+        _ebpf_pining_table_extract);
```
Comment (dthaler) [inline libs/platform/ebpf_pinning_table.c:56]:
```suggestion
        _ebpf_pinning_table_extract);
```

File: libs/platform/ebpf_platform.h [hunk 14]
```diff
@@ -402,7 +397,8 @@ extern "C"
         _In_ void (*free)(void* memory),
         size_t key_size,
         size_t value_size,
-        _In_opt_ ebpf_hash_table_compare_result_t (*compare_function)(const uint8_t* key1, const uint8_t* key2));
+        size_t bucket_count,
+        void (*extract_function)(_In_ const uint8_t* value, _Out_ const uint8_t** data, _Out_ size_t* length));
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:401]:
```suggestion
        _In_ void (*extract_function)(_In_ const uint8_t* value, _Outptr_ const uint8_t** data, _Out_ size_t* length));
```

File: libs/platform/ebpf_platform.h [hunk 15]
```diff
@@ -547,6 +543,25 @@ extern "C"
...
+     *  if the input value pointed to by destination equals the value of
+     *  comperand.
+     * @param[in] comperand Specifies the value that is compared with the input
+     *  value pointed to by destination.
+     * @return Returns the original value of memory pointed to by
+     *  destination.
+     */
+    void*
+    ebpf_interlocked_compare_exchange_pointer(
+        _Inout_ void* volatile* destination, const void* exchange, const void* comperand);
```
Comment (dthaler) [inline libs/platform/ebpf_platform.h:563]:
```suggestion
        _Inout_ void* volatile* destination, _In_ const void* exchange, _In_ const void* comperand);
```

File: libs/platform/ebpf_hash_table.c [hunk 16]
```diff
@@ -3,52 +3,321 @@
...
 #include "ebpf_platform.h"
 
+// Buckets contain an array of pointers to data blocks containing key + value.
+// Buckets are immutable once inserted in to the hash-table and replaced when
+// modified.
+
+// Layout is:
+// ebpf_hash_table_t.buckets->ebpf_hash_bucket_header_t.entries->data
+// Keys are stored contiguously in ebpf_hash_bucket_header_t for fast
+// searching, data is stored separately to prevent RCU from causing loss of
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:13]: please expand RCU

File: libs/platform/ebpf_hash_table.c [hunk 17]
```diff
@@ -3,52 +3,321 @@
...
+typedef struct _ebpf_hash_bucket_entry
+{
+    uint8_t* data;
+    uint8_t key[1];
+} ebpf_hash_bucket_entry_t;
+
+typedef struct _ebpf_hash_bucket_header
+{
+    size_t count;
+    ebpf_hash_bucket_entry_t entries[1];
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:25]:
```suggestion
    _Field_size_(count)
    ebpf_hash_bucket_entry_t entries[1];
```
(see https://docs.microsoft.com/en-us/cpp/code-quality/annotating-structs-and-classes?view=msvc-160 for explanation of this suggestion)

File: libs/platform/ebpf_hash_table.c [hunk 18]
```diff
@@ -3,52 +3,321 @@
...
-    ebpf_hash_table_compare_result_t (*compare_function)(const uint8_t* key1, const uint8_t* key2);
+    uint32_t bucket_count;
+    volatile int32_t entry_count;
+    uint32_t seed;
     size_t key_size;
     size_t value_size;
     void* (*allocate)(size_t size);
     void (*free)(void* memory);
+    void (*extract)(const uint8_t* value, const uint8_t** data, size_t* num);
+    ebpf_hash_bucket_header_t* volatile buckets[1];
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:38]:
```suggestion
    _Field_size_(bucket_count) 
    ebpf_hash_bucket_header_t* volatile buckets[1];
```

File: libs/platform/ebpf_hash_table.c [hunk 19]
```diff
@@ -3,52 +3,321 @@
...
-    if (table->compare_function) {
-        return (RTL_GENERIC_COMPARE_RESULTS)table->compare_function(
-            (const uint8_t*)first_struct, (const uint8_t*)second_struct);
+/**
+ * @brief Given two potentially non-comparable key values, extract the key and
+ * compare them.
+ *
+ * @param[in] hash_table Hash table the keys belong to.
+ * @param[in] key_a First key
+ * @param[in] key_b Second key
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:119]:
```suggestion
 * @param[in] key_a First key.
 * @param[in] key_b Second key.
```

File: libs/platform/ebpf_hash_table.c [hunk 20]
```diff
@@ -3,52 +3,321 @@
...
+ * @retval -1 if key_a < key_b
+ * @retval 0 if key_a == key_b
+ * @retval 1 if key_a > key_b
+ */
+static int
+_ebpf_hash_table_compare(ebpf_hash_table_t* hash_table, _In_ const uint8_t* key_a, _In_ const uint8_t* key_b)
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:125]: add SAL annotation to hash_table argument.  Is it `_Inout_` or should it be `_In_ const ebpf_hash_table_t*`?

File: libs/platform/ebpf_hash_table.c [hunk 21]
```diff
@@ -3,52 +3,344 @@
...
 }
 
-static void*
-_ebpf_hash_table_allocate(struct _RTL_AVL_TABLE* avl_table, unsigned long byte_size)
+/**
+ * @brief Given a potentially non-comparable key value, extract the key and
+ * compute the hash.
+ *
+ * @param[in] hash_table Hash table the keys belong to.
+ * @param[in] key_a First key
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:154]: "First" in what sense?  I suspect this description is a cut and paste error :)

File: libs/platform/ebpf_hash_table.c [hunk 22]
```diff
@@ -3,52 +3,344 @@
...
+ * @retval Hash of key
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:155]:
```suggestion
 * @return Hash of key
```
(only use retval for discrete values)

File: libs/platform/ebpf_hash_table.c [hunk 23]
```diff
@@ -3,52 +3,344 @@
...
+ */
+static uint32_t
+_ebpf_hash_table_compute_hash(ebpf_hash_table_t* hash_table, _In_ const uint8_t* key)
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:158]: add SAL annotation for hash_table argument

File: libs/platform/ebpf_hash_table.c [hunk 24]
```diff
@@ -3,52 +3,344 @@
...
 }
 
-static void
-_ebpf_hash_table_free(struct _RTL_AVL_TABLE* avl_table, void* buffer)
+/**
+ * @brief Given a pointer to a bucket, compute the offset of a bucket entry.
+ *
+ * @param [in] key_size Size
+ * @param [in] bucket Pointer to start of the bucket.
+ * @param [in] index Index into the bucket
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:176]:
```suggestion
 * @param [in] key_size Size.
 * @param [in] bucket Pointer to start of the bucket.
 * @param [in] index Index into the bucket.
```
nit: use consistent punctuation

File: libs/platform/ebpf_hash_table.c [hunk 25]
```diff
@@ -3,52 +3,344 @@
...
+    for (;;) {
+        size_t old_bucket_count = 0;
+        size_t new_bucket_count = 0;
+
+        new_bucket = NULL;
+        old_data = NULL;
+        delete_bucket = new_bucket;
+        delete_data = new_data;
+
+        // Capture current bucket pointer
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:241]:
```suggestion
        // Capture current bucket pointer.
```

File: libs/platform/ebpf_hash_table.c [hunk 26]
```diff
@@ -3,52 +3,344 @@
...
+        old_bucket = hash_table->buckets[hash % hash_table->bucket_count];
+
+        // If the old_bucket exists, capture it's count.
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:244]:
```suggestion
        // If the old_bucket exists, capture its count.
```

File: libs/platform/ebpf_hash_table.c [hunk 27]
```diff
@@ -92,144 +390,168 @@ ebpf_hash_table_create(
...
-            // Advance the cursor until we reach the first entry that is greater than
-            // the key.
-            while (!(_ebpf_hash_map_compare(table, (uint8_t*)previous_key, entry) == GenericGreaterThan)) {
-                entry = RtlEnumerateGenericTableAvl(table, FALSE);
-                if (!entry) {
-                    break;
-                }
+            // Is this the previous key?
+            if (_ebpf_hash_table_compare(hash_table, previous_key, entry->key) == 0) {
+                // Yes, record it's location.
```
Comment (dthaler) [inline libs/platform/ebpf_hash_table.c:533]:
```suggestion
                // Yes, record its location.
```

File: libs/platform/ebpf_hash_table.c (hunk 11 above)
Comment (Alan-Jowett) [reply libs/platform/ebpf_hash_table.c]: Add it when and if the need arises.

File: libs/execution_context/ebpf_program.c [hunk 28]
```diff
@@ -249,9 +249,6 @@ ebpf_program_create(ebpf_program_t** program)
         goto Done;
     }
 
-    ebpf_list_initialize(&local_program->links);
-    ebpf_lock_create(&local_program->links_lock);
```
Comment (dthaler) [inline libs/execution_context/ebpf_program.c:253]: This change looks like it's reverting PR #395.  Is that intentional?

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_program.c]: No, not intentional

**Review by dthaler — APPROVED** [review]
